lib/index.js

```diff
@@ -99,366 +96,295 @@ exports.Boom = class extends Error {
 
         return this.prototype.isPrototypeOf(instance);
     }
-};
-
 
-exports.isBoom = function (err, statusCode) {
+    reformat(debug = false) {
 
-    return err instanceof Error && !!err.isBoom && (!statusCode || err.output.statusCode === statusCode);
-};
+        this.output.payload = new internals.PayloadObject(this, this.output.statusCode, debug);
+    }
 
+    #apply(data, statusCode, headers, message) {
 
-exports.boomify = function (err, options) {
+        if (data !== undefined) {
+            this.data = data;
+        }
 
-    Hoek.assert(err instanceof Error, 'Cannot wrap non-Error object');
+        if (statusCode) {
+            const numberCode = parseInt(statusCode, 10);
+            if (isNaN(numberCode) || numberCode < 400) {
+                throw new TypeError(`statusCode must be a number (400+): ${statusCode}`);
+            }
 
-    options = options || {};
+            if (message) {
+                this.message = `${message}: ${this.message}`;
+            }
 
-    if (options.data !== undefined) {
-        err.data = options.data;
+            const payload = new internals.PayloadObject(this, numberCode, false);
+            this.output = new internals.BoomOutput(numberCode, payload, headers);
+        }
     }
 
-    if (options.decorate) {
-        Object.assign(err, options.decorate);
-    }
+    static {
+        Object.defineProperty(this.prototype, 'name', { value: 'Boom', writable: true, configurable: true });
+        Object.defineProperty(this.prototype, 'isBoom', { value: true, configurable: true });
+
+        exports.isBoom = function (err, statusCode) {
+
+            return err instanceof Error && !!err.isBoom && (!statusCode || err.output.statusCode === statusCode);
+        };
+
+        exports.boomify = function (err, options = {}) {
+
+            const { override, data, statusCode, message } = options;
+
+            if (!err?.isBoom) {
+                return new exports.Boom(message, { statusCode, cause: err, data });
+            }
 
-    if (!err.isBoom) {
-        return internals.initialize(err, options.statusCode ?? 500, options.message);
+            if (override === false) {                                 // Defaults to true
+                err.#apply(data);
+            }
+            else {
+                err.#apply(data, statusCode ?? err.output.statusCode, {}, message);
+            }
```

Comment:
You might want to use `instanceof` here, otherwise errors from older version of this lib might have `err.isBoom` equal to `true`, but `err.#apply` will not be allowed. Same goes if this lib (even in its current version) get bundled twice.

Reply:
Actually, you might want to test `boomify()` with older versions of the lib more thoroughly !

Reply:
Good catch! I forgot that `boomify()` will be expected to boomify legacy boom objects.

I have updated the code to not rely on private fields, and created actual tests to confirm legacy support.